src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.cs

```diff
@@ -0,0 +1,252 @@
+            {
+                return wasUpdated;
+            }
+
+            var activeConfiguredProject = await GetActiveConfiguredProjectByPathAsync(projectPath);
+
+            var referenceHandler = FindReferenceHandler(referenceUpdate);
+
+            wasUpdated = await ApplyAction(referenceUpdate, cancellationToken, referenceHandler, activeConfiguredProject);
+
+            return wasUpdated;
+        }
+
+        private static async Task<bool> ApplyAction(ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken,
+            ReferenceHandler referenceHandler, ConfiguredProject activeConfiguredProject)
+        {
+            bool wasUpdated;
+
+            if (referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUsed ||
+                referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUnused)
+            {
+                wasUpdated =
+                    await referenceHandler.UpdateReferenceAsync(activeConfiguredProject, referenceUpdate, cancellationToken);
+            }
+            else
+            {
+                if (!canRemoveReference(referenceUpdate, referenceHandler))
+                {
+                    wasUpdated = false;
+                }
+                else
+                {
+                    wasUpdated =
+                        await referenceHandler.RemoveReferenceAsync(activeConfiguredProject, referenceUpdate.ReferenceInfo);
+                }
+            }
+
+            return wasUpdated;
+        }
+
+        private static bool canRemoveReference(ProjectSystemReferenceUpdate referenceUpdate, ReferenceHandler referenceHandler)
+        {
+            var references = referenceHandler.GetReferences();
+
+            var reference = references.Where(c =>
+                c.ItemSpecification == referenceUpdate.ReferenceInfo.ItemSpecification).First();
+
+            if (reference.TreatAsUsed == true)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private ReferenceHandler FindReferenceHandler(ProjectSystemReferenceUpdate referenceUpdate)
+        {
+            ReferenceHandler referenceHandler;
+
+            if (referenceUpdate.ReferenceInfo.ReferenceType == ProjectSystemReferenceType.Project)
+            {
+                referenceHandler = _projectReferenceHandler;
+            }
+            else if (referenceUpdate.ReferenceInfo.ReferenceType == ProjectSystemReferenceType.Package)
+            {
+                referenceHandler = _packageReferenceHandler;
+            }
+            else if (referenceUpdate.ReferenceInfo.ReferenceType == ProjectSystemReferenceType.Assembly)
+            {
+                referenceHandler = _assemblyReferenceHandler;
+            }
+            else
+            {
+                referenceHandler = _sdkReferenceHandler;
+            }
+
+            return referenceHandler;
+        }
+
+        public bool IsProjectCpsBased(string projectPath)
```

Comment:
Where is this API used? I can't see it as part of the API definition. We avoid using "Cps" in public APIs.

Reply:
removed, it was to differentiate zero references in a non-cps project vs a project with no references.